cgen error, for pure V program, overloading `<` for structs, and calling `arrays.distinct` with an array of structs, that sorts them internally

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 dab25ca.43e27d2
OS: linux, Ubuntu 20.04.6 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /space/v/oo
vexe: /space/v/oo/v
vexe mtime: 2024-12-10 11:52:21

vroot: OK, value: /space/v/oo
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.0
Git vroot status: weekly.2024.50-3-g43e27d24
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg /home/delian/code/misc/2024_12_10__13/cgen_error.v`

```v
import arrays
struct Abc { x int }
fn (s1 Abc) < (s2 Abc) bool { return s1.x < s2.x }
a := [Abc{123}, Abc{42}, Abc{999}, Abc{42}, Abc{123}]
dump(arrays.distinct(a))
```

**What did you expect to see?**

A compiled program

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/cgen_error.01JER6KR8W4G6C6194G79DHWTH.tmp.c:1623: warning: implicit declaration of function 'main__Abc__lt'
cc: /tmp/v_1000/cgen_error.01JER6KR8W4G6C6194G79DHWTH.tmp.c:1627: error: incompatible types for redefinition of 'main__Abc__lt'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzU4MmJmYTFlMDYzMmRkMDhlMmFlYzQiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.JJ37d9b26NOY7ptJWHBYhIFiJKKYMUhJItBVY1YcamI">Huly&reg;: <b>V_0.6-21557</b></a></sub>

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -766,7 +766,7 @@ fn (mut g Gen) gen_array_sort(node ast.CallExpr) {
 	}
 
 	stype_arg := g.styp(elem_type)
-	g.definitions.writeln('VV_LOCAL_SYMBOL ${g.static_modifier} int ${compare_fn}(${stype_arg}* a, ${stype_arg}* b) {')
+	g.sort_fn_definitions.writeln('VV_LOCAL_SYMBOL ${g.static_modifier} int ${compare_fn}(${stype_arg}* a, ${stype_arg}* b) {')
 	c_condition := if comparison_type.sym.has_method('<') {
 		'${g.styp(comparison_type.typ)}__lt(${left_expr}, ${right_expr})'
 	} else if comparison_type.unaliased_sym.has_method('<') {
@@ -776,9 +776,9 @@ fn (mut g Gen) gen_array_sort(node ast.CallExpr) {
 	} else {
 		'${left_expr} < ${right_expr}'
 	}
-	g.definitions.writeln('\tif (${c_condition}) return -1;')
-	g.definitions.writeln('\telse return 1;')
-	g.definitions.writeln('}\n')
+	g.sort_fn_definitions.writeln('\tif (${c_condition}) return -1;')
+	g.sort_fn_definitions.writeln('\telse return 1;')
+	g.sort_fn_definitions.writeln('}\n')
 
 	// write call to the generated function
 	g.gen_array_sort_call(node, compare_fn, left_is_array)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -59,6 +59,7 @@ mut:
 	enum_typedefs             strings.Builder // enum types
 	definitions               strings.Builder // typedefs, defines etc (everything that goes to the top of the file)
 	type_definitions          strings.Builder // typedefs, defines etc (everything that goes to the top of the file)
+	sort_fn_definitions       strings.Builder // sort fns
 	alias_definitions         strings.Builder // alias fixed array of non-builtin
 	hotcode_definitions       strings.Builder // -live declarations & functions
 	channel_definitions       strings.Builder // channel related code
@@ -304,6 +305,7 @@ pub fn gen(files []&ast.File, mut table ast.Table, pref_ &pref.Preferences) (str
 		typedefs:             strings.new_builder(100)
 		enum_typedefs:        strings.new_builder(100)
 		type_definitions:     strings.new_builder(100)
+		sort_fn_definitions:  strings.new_builder(100)
 		alias_definitions:    strings.new_builder(100)
 		hotcode_definitions:  strings.new_builder(100)
 		channel_definitions:  strings.new_builder(100)
@@ -392,6 +394,7 @@ pub fn gen(files []&ast.File, mut table ast.Table, pref_ &pref.Preferences) (str
 			global_g.includes.write(g.includes) or { panic(err) }
 			global_g.typedefs.write(g.typedefs) or { panic(err) }
 			global_g.type_definitions.write(g.type_definitions) or { panic(err) }
+			global_g.sort_fn_definitions.write(g.sort_fn_definitions) or { panic(err) }
 			global_g.alias_definitions.write(g.alias_definitions) or { panic(err) }
 			global_g.definitions.write(g.definitions) or { panic(err) }
 			global_g.gowrappers.write(g.gowrappers) or { panic(err) }
@@ -590,6 +593,7 @@ pub fn gen(files []&ast.File, mut table ast.Table, pref_ &pref.Preferences) (str
 	b.write_string2('\n// V result_xxx definitions:\n', g.out_results.str())
 	b.write_string2('\n// V json forward decls:\n', g.json_forward_decls.str())
 	b.write_string2('\n// V definitions:\n', g.definitions.str())
+	b.write_string2('\n// V sort fn definitions:\n', g.sort_fn_definitions.str())
 	b.writeln('\n// V global/const non-precomputed definitions:')
 	for var_name in g.sorted_global_const_names {
 		if var := g.global_const_defs[var_name] {
@@ -682,6 +686,7 @@ fn cgen_process_one_file_cb(mut p pool.PoolProcessor, idx int, wid int) &Gen {
 		includes:              strings.new_builder(100)
 		typedefs:              strings.new_builder(100)
 		type_definitions:      strings.new_builder(100)
+		sort_fn_definitions:   strings.new_builder(100)
 		alias_definitions:     strings.new_builder(100)
 		definitions:           strings.new_builder(100)
 		gowrappers:            strings.new_builder(100)
@@ -754,6 +759,7 @@ pub fn (mut g Gen) free_builders() {
 		g.includes.free()
 		g.typedefs.free()
 		g.type_definitions.free()
+		g.sort_fn_definitions.free()
 		g.alias_definitions.free()
 		g.definitions.free()
 		g.cleanup.free()

--- a/vlib/v/tests/sort_fn_def_test.v
+++ b/vlib/v/tests/sort_fn_def_test.v
@@ -1,0 +1,15 @@
+import arrays
+
+struct Abc {
+	x int
+}
+
+fn (s1 Abc) < (s2 Abc) bool {
+	return s1.x < s2.x
+}
+
+fn test_main() {
+	a := [Abc{123}, Abc{42}, Abc{999}, Abc{42}, Abc{123}]
+	dump(arrays.distinct(a))
+	assert true
+}

